test-gvfs-prococol, t5799: tests for gvfs-helper

jeffhostetler opened the pull request:
**This PR is based on 2.24.0 and replaces PR https://github.com/microsoft/git/pull/216
which was based on 2.23.0.**

Create t/helper/test-gvfs-protocol.c and t/t5799-gvfs-helper.sh
to test gvfs-helper.

Create t/helper/test-gvfs-protocol.c as a stand-alone web server that
speaks the GVFS Protocol [1] and serves loose objects and packfiles
to clients.  It is borrows heavily from the code in daemon.c.
It includes a "mayhem" mode to cause various network and HTTP errors
to test the retry/recovery ability of gvfs-helper.

Create t/t5799-gvfs-helper.sh to test gvfs-helper.

jeffhostetler committed:
test-gvfs-prococol, t5799: tests for gvfs-helper

Create t/helper/test-gvfs-protocol.c and t/t5799-gvfs-helper.sh
to test gvfs-helper.

Create t/helper/test-gvfs-protocol.c as a stand-alone web server that
speaks the GVFS Protocol [1] and serves loose objects and packfiles
to clients.  It is borrows heavily from the code in daemon.c.
It includes a "mayhem" mode to cause various network and HTTP errors
to test the retry/recovery ability of gvfs-helper.

Create t/t5799-gvfs-helper.sh to test gvfs-helper.

[1] https://github.com/microsoft/VFSForGit/blob/master/Protocol.md

Signed-off-by: Jeff Hostetler <jeffhost@microsoft.com>

derrickstolee commented:
> It includes a "mayhem" mode to cause various network and HTTP errors
> to test the retry/recovery ability of gvfs-helper.

Mayhem mode sounds so much more fun than ["chaos"](https://en.wikipedia.org/wiki/Chaos_engineering).

derrickstolee commented:
@jeffhostetler I'm sure you're on top of this, but your builds seem to be failing because `test-gvfs-helper` is not deleting properly, possibly because it is still running?

dscho commented:
> builds seem to be failing because `test-gvfs-helper` is not deleting properly

That's not actually the problem, the problem is that `t/helper/test-gvfs-helper.exe` is built, but it is not added to the `.gitignore` file. (The [error message](https://gvfs.visualstudio.com/ci/_build/results?buildId=17535&view=logs&j=8b03b7e8-2f7a-5d24-990c-fbca7153226b&t=581b985a-e632-5387-73c5-f8a7ad819b78&l=1955) was supposed to be indicative of that.)

dscho commented on t/t5799-gvfs-helper.sh:
```diff
@@ -0,0 +1,972 @@
+		actual_nr=$(( $(wc -l <"$OIDS_FILE") ))
+		if test $actual_nr -lt $1
+		then
+			echo "get_list_of_oids: insufficient data.  Need $1 OIDs."
```
... `>&2` ;-)

jeffhostetler replied:
@dscho  where do these messages appear (whether they come out on 1 or 2) ?

that is, where are stdout and stderr mapped to for each of these script functions and the test_expect functions ?

dscho replied:
By default, all output is suppressed. When running the test scripts with the `-v` (or `--verbose`) option, the output is no longer suppressed. The only difference, then, between `stdout` and `stderr` is the buffering: messages sent to `stderr` appear earlier (which matters when running with `-x` because then `stdout` might mislead the reader into believing that a different command produced that output).

As a rule, error messages should go to `stderr`.

Not a big deal, though. A lot of test scripts in Git's test suite get this wrong.

dscho commented on t/t5799-gvfs-helper.sh:
```diff
@@ -0,0 +1,972 @@
+
+	if test $# -eq 1
+	then
+		actual_nr=$(( $(wc -l <"$OIDS_FILE") ))
```
The idiomatic way to do this in Git's test suite is `actual_nr=$(wc -l "$OIDS_FILE")`.

jeffhostetler replied:
I had several problems getting this function to work as I wanted and settled on this.
Let me ask a few questions about the problems that I fought thru to get this working.

When I do `x=$(wc -l file)`, I get the count and the filename in the output.

When I do `x=$(wc -l <file)`, I just get the count in the output.  This saves me from having to run sed/awk on the output, right?

On the Mac, `wc -l` yields a string with leading spaces.  This doesn't play right with the `-lt` or `-eq` operators.  Wrapping the expression in `$(( ... ))` gives me a number which behaved correctly.

Does the variable assignment eat the leading whitespace, such that `x=$(wc -l <file)` is the right answer ??

dscho replied:
> When I do `x=$(wc -l file)`, I get the count and the filename in the output.

I forgot about that, sorry.

> When I do `x=$(wc -l <file)`, I just get the count in the output. This saves me from having to run sed/awk on the output, right?

Yes.

> On the Mac, `wc -l` yields a string with leading spaces. This doesn't play right with the `-lt` or `-eq` operators. Wrapping the expression in `$(( ... ))` gives me a number which behaved correctly.

Right, but if you do not wrap `$actual_nr` in double-quotes, the string is white-space trimmed implictly, too.

> Does the variable assignment eat the leading whitespace, such that `x=$(wc -l <file)` is the right answer ??

Not quite. The quoting is not necessary here, but I think it will still reduce the whitespace to a single space on macOS. Leaving off the double-quotes when expanding `$actual_nr` is the actual trick.

dscho commented on t/t5799-gvfs-helper.sh:
```diff
@@ -0,0 +1,972 @@
+	cp "$REPO_SRC"/.git/refs/heads/master m2.branch &&
+	#
+	# test_commit() creates commits, trees, tags, and blobs and leave
+	# them loose.
```
You might want to force auto-gc off via `test_config gc.auto 0 &&` so that this promise is definitely kept.

derrickstolee replied:
But we will be very unlikely to surpass the default limit for triggering a GC (but who knows what happens in the future).

dscho replied:
True.

dscho commented on t/t5799-gvfs-helper.sh:
```diff
@@ -0,0 +1,972 @@
+		expected_nr=1
+	fi
+
+	actual_nr=$(( $(grep "packfile " OUT.output | wc -l) ))
```
How about `actual_nr=$(grep -c "packfile " <OUT.output`?

jeffhostetler replied:
Cool.  A new option I didn't know about.

dscho replied:
:-) TBH I had not known about it until some time earlier this year, I think. In any case, I still always have to look at https://man.openbsd.org/grep and https://linux.die.net/man/1/grep and sometimes at https://pubs.opengroup.org/onlinepubs/009695399/utilities/grep.html to figure out what is portable, as GNU grep's man pages seem to delight in omitting the information what options are standard compliant and what options are GNU extensions.

dscho reviewed (commented):
I cannot really comment on the GVFS-related internals, of course, so I just left a couple of generic comments ;-)

jeffhostetler committed:
t/helper/.gitignore: add test-gvfs-procotol

Signed-off-by: Jeff Hostetler <jeffhost@microsoft.com>

jeffhostetler commented:
WRT the test cleanup failures: Tricky, there's a t/helper/.gitignore file.  Pushing a fix for that.

jeffhostetler committed:
gvfs-helper: move result-list construction into install functions

gvfs-helper prints a "loose <oid>" or "packfile <name>" messages after
they are received to help invokers update their in-memory caches.
Move the code to accumulate these messages in the result_list into
the install_* functions rather than waiting until the end.

POST requests containing 1 object may return a loose object or a packfile
depending on whether the object is a commit or non-commit.  Delaying the
message generation just complicated the caller.

Signed-off-by: Jeff Hostetler <jeffhost@microsoft.com>

jeffhostetler committed:
t5799: add support for POST to return either a loose object or packfile

Earlier versions of the test always returned a packfile in response to a POST.
Now we look at the number of objects in the POST request.

If > 1, always send a packfile.

If = 1 and it is a commit, send a packfile.
Otherwise, send a loose object.

This is to better model the behavior of the GVFS server/protocol which
treats commits differently.

Signed-off-by: Jeff Hostetler <jeffhost@microsoft.com>

jeffhostetler commented:
@derrickstolee   I rebased this branch on the current tip so that it includes your fix for POST and commits.  Added a few fixes for that and included unit tests to confirm that we properly receive either packfiles or a loose object when requesting a single object.

jeffhostetler committed:
t5799: cleanup wc-l and grep-c lines

Signed-off-by: Jeff Hostetler <jeffhost@microsoft.com>

jeffhostetler commented:
I've addressed the "wc -l" and "grep -c" suggestions.  I'm going to leave the stderr redirect for another day.

dscho commented:
> I've addressed the "wc -l" and "grep -c" suggestions. I'm going to leave the stderr redirect for another day.

Sounds good!
